add a section about memory oversubscription (#10573)
hashicorp/nomad: website/content/docs/job-specification/resources.mdx

- `memory` `(int: 300)` - Specifies the memory required in MB.

> - `memory_max` <code>(`int`: &lt;optional&gt;)</code> <sup>1.1 Beta</sup> - Optionally, specifies the maximum memory the task may use, if the client has excess memory capacity, in MB. See the [Memory Oversubscription for more details](#memory-oversubscription).

Reviewer: I'd have the link just around the section title here:

```suggestion
- `memory_max` <code>(`int`: &lt;optional&gt;)</code> <sup>1.1 Beta</sup> - Optionally, specifies the maximum memory the task may use, if the client has excess memory capacity, in MB. See [Memory Oversubscription](#memory-oversubscription) for more details.
```

  }
}
```
## Memory Oversubscription

> Setting task memory limits is a challenging act of balance. Job authors must set
> a memory limit for each task — if it’s too low, the task may be interrupted if
> it exceeds the limit unexpectedly; if it’s too high, the cluster is left
> underutilized and resources are wasted. Job authors usually set limits based on
> their task's typical memory usage and then add a safety margin to handle
> unexpected load spikes or uncommon scenarios. That means that in some clusters,
> 20-40% of the cluster memory may be reserved but unused.
> 
> To help minimize reserved but not used cluster memory, Nomad 1.1 lets job
> authors set two separate memory limits:

Reviewer: Because we took this from the blog post it's reading as a bit "marketing-y" and it includes operator behaviors we're looking to mitigate, rather than focusing on . I'd tighten this up as follows:

```suggestion
Setting task memory limits requires balancing the risk of interrupting tasks
against the risk of wasting resources. If a task memory limit is set too low,
the task may exceed the limit and be interrupted; if the task memory is too
high, the cluster is left underutilized.

To help maximize cluster memory utilization while allowing a safety margin for
unexpected load spikes, Nomad 1.1. lets job authors set two separate memory
limits:
```

* `memory`: the reserve limit to represent the task’s typical memory usage —
  this number is used by the Nomad scheduler to reserve and place the task

* `memory_max`: the maximum memory the task may use, if the client has excess
  available memory, and may be terminated if it exceeds

> If a client's memory becomes contended or low, the operating system will
> pressure the running tasks to free up memory. If the contention persist, Nomad
> may kill oversubscriped tasks and reschedule them to other clients. The exact
> mechanism for memory pressure is specific to the task driver and application
> runtime.

Reviewer: ```suggestion
If a client's memory becomes contended or low, the operating system will
pressure the running tasks to free up memory. If the contention persists, Nomad
may kill oversubscribed tasks and reschedule them to other clients. The exact
mechanism for memory pressure is specific to the task driver, operating system,
and application runtime.
```

> The new max limit attribute is currently supported by Docker and all default
> Nomad drivers except raw_exec and QEMU. Consult the third-party drivers
> documentation for memory oversubscription support.

Reviewer: Let's make the supported driver list explicit. Making the list explicit also makes me ask if `podman` has support? It's neither "default" nor "third-party"

```suggestion
The new max limit attribute is currently supported by the official `docker`,
`exec`, and `java` task drivers.  Consult the documentation of
community-supported task drivers for their memory oversubscription support.
```

> Nomad operators must enable [Memory Oversubscription in Scheduler
> Config](/api-docs/operator/scheduler#update-scheduler-configuration); memory
> oversubscription is opt-in. Enterprise customers can use [Resource

Reviewer: "Opt-in" is the most important part of this paragraph, so if we make it the first sentence that'll pop out to readers.

```suggestion
Memory oversubscription is opt-in. Nomad operators can enable [Memory Oversubscription in the scheduler
configuration](/api-docs/operator/scheduler#update-scheduler-configuration). Enterprise customers can use [Resource
```

> Quotas](https://learn.hashicorp.com/tutorials/nomad/quotas) to limit the memory

Author: TODO: Document the quotas for memory oversubscription.

oversubscription.

To avoid degrading the cluster experience, we recommend examining and monitoring
resource utilization and considering the following suggestions:

* Set `oom_score_adj` for Linux host services that aren't managed by Nomad, e.g.
  Docker, logging services, and the Nomad agent itself. For Systemd services, you can use the [`OOMScoreAdj` field](https://github.com/hashicorp/nomad/blob/v1.0.0/dist/systemd/nomad.service#L25).

* Monitor hosts for memory utilization and set alerts on Out-Of-Memory errors

> * Set the [client `reserved`](/docs/configuration/client#reserved) with enough memory for host unmanaged services as well as a buffer for the memory excess. e.g. If the client reserved memory is 1GB, the allocations on the host may exceeed their soft memory limit by almost 1GB in aggregate before the memory becomes contended and allocations get killed.

Reviewer: The use of "e.g." here is technically correct but if you've got a full sentence of stuff after it, it reads nicer to make that a standalone sentence. Also cleaned up the "host unmanaged services" for clarity.

```suggestion
* Set the [client `reserved`](/docs/configuration/client#reserved) with enough
  memory for host services that aren't managed by Nomad as well as a buffer
  for the memory excess. For example, if the client reserved memory is 1GB,
  the allocations on the host may exceed their soft memory limit by almost
  1GB in aggregate before the memory becomes contended and allocations get
  killed.
```

  }
}
```

> ## Memory Oversubscription

Author: FWIW, Not sure where this belongs. This seems the most relevant part - though, I considered having it in Internal/Scheduling documents.

Reviewer: This location seems fine.
